lib/Twig/Environment.php

```diff
@@ -314,11 +314,15 @@ public function loadTemplate($name, $index = null)
         if (!class_exists($cls, false)) {
             $key = $this->cache->generateKey($name, $cls);
 
-            if (!$this->cache->has($key) || ($this->isAutoReload() && !$this->isTemplateFresh($name, $this->cache->getTimestamp($key)))) {
-                $this->cache->write($key, $this->compileSource($this->getLoader()->getSource($name), $name));
+            if (!$this->isAutoReload() || $this->isTemplateFresh($name, $this->cache->getTimestamp($key))) {
+                $this->cache->load($key);
             }
 
-            $this->cache->load($key);
+            if (!class_exists($cls, false)) {
+                $content = $this->compileSource($this->getLoader()->getSource($name), $name);
+                $this->cache->write($key, $content);
+                eval('?>'.$content);
```

Comment:
I agree with @znerol that this is the correct approach. And it should not have a performance impact as this eval is only happening if the cache doesn't exist. And this this case (a new generated class), the opcode caching wouldn't happen anyway.